### PR TITLE
Add kafkaMurmurHash function

### LINKED ISSUE #47834: New murmur2 function (use Kafka DefaultPartitioner seed value)
As we are consuming from Kafka using the DefaultPartitioner as the partition strategy and writing into an `all-sharded` cluster, it would be a good idea to have the murmur2 function that is implement by Kafka using the specific seed value so that we could implement a `Distributed Table` with such a hashing functionality and query for the specific key directly on the resolving cluster node.

Seed information can be found here: https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L480
```
    @SuppressWarnings("fallthrough")
    public static int murmur2(final byte[] data) {
        int length = data.length;
        int seed = 0x9747b28c;
        // 'm' and 'r' are mixing constants generated offline.
        // They're not really 'magic', they just happen to work well.
        final int m = 0x5bd1e995;
        final int r = 24;
```

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add kafkaMurmurHash function for compatibility with Kafka DefaultPartitioner. Closes https://github.com/ClickHouse/ClickHouse/issues/47834

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by evillique** [comment]
Test failures:
- AST fuzzer (debug): https://github.com/ClickHouse/ClickHouse/issues/48233
- AST fuzzer (ubsan): https://github.com/ClickHouse/ClickHouse/issues/48053
- Integration tests (asan) [1/6]: https://github.com/ClickHouse/ClickHouse/issues/48180
- Integration tests (tsan) [1/6]: https://github.com/ClickHouse/ClickHouse/issues/48181

File: src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -465,6 +465,26 @@ struct GccMurmurHashImpl
...
+/// To be compatible with Kafka: https://github.com/apache/kafka/blob/461c5cfe056db0951d9b74f5adc45973670404d7/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L480
+struct KafkaMurmurHashImpl
+{
+    static constexpr auto name = "kafkaMurmurHash";
+
+    using ReturnType = Int32;
+
+    static Int32 apply(const char * data, const size_t size)
+    {
+        return static_cast<ReturnType>(MurmurHash2(data, size, 0x9747b28cU));
```
Comment (dbug-river) [inline src/Functions/FunctionsHashing.h:477]: The kafka murmurhash always returns a positive numerical value, I would suggest something like this

`    
    using ReturnType = UInt32;

    static UInt32 apply(const char * data, const size_t size)
    {
        return MurmurHash2(data, size, 0x9747b28cUL) & 0x7fffffff;
    }
`

Comment (evillique) [reply src/Functions/FunctionsHashing.h:477]: Are the unit tests to Murmur in the Kafka repository hash wrong then?https://github.com/apache/kafka/blob/139f7709bd3f5926901a21e55043388728ccca78/clients/src/test/java/org/apache/kafka/common/utils/UtilsTest.java#L93
Or how is it possible to check the results of this function from inside Kafka?

Comment (evillique) [reply src/Functions/FunctionsHashing.h:477]: I specifically changed the return type to correspond with Kafka's version that returns Int32.

Comment (evillique) [reply src/Functions/FunctionsHashing.h:477]: Yep, though original murmur2 function returns Int32, it is automatically casted to UInt32 in DefaultPartitioner: https://github.com/apache/kafka/blob/139f7709bd3f5926901a21e55043388728ccca78/clients/src/main/java/org/apache/kafka/clients/producer/internals/BuiltInPartitioner.java#L328

**Review by Avogar — APPROVED** [review]

**Pull request merged**